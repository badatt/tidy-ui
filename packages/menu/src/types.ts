import React, { HTMLAttributes } from 'react';
import { IBaseProps, TTone } from '@tidy-ui/types';

/**
 * Props for the Menu component.
 *
 * The Menu component provides a container for menu items with customizable styling
 * and trigger elements.
 *
 * @interface IMenuProps
 */
export interface IMenuProps extends IBaseProps, HTMLAttributes<HTMLMenuElement> {
  /**
   * Controls whether the trigger element is disabled, preventing it from opening the menu.
   *
   * @type {boolean}
   * @default false
   */
  disabledTrigger?: boolean;

  /**
   * Controls whether the menu is open or closed.
   *
   * @type {boolean}
   * @default false
   */
  isOpen?: boolean;

  /**
   * Controls whether the menu has sharp (squared) corners or rounded corners.
   * When true, the menu will display with sharp corners instead of the default rounded appearance.
   *
   * @type {boolean}
   * @default false
   */
  isSharp?: boolean;

  /**
   * The trigger element that will open/close the menu when interacted with.
   * This should be a React element that can receive HTML attributes (like buttons, divs, etc.).
   * The trigger element will typically handle click events to show/hide the menu.
   *
   * @type {React.ReactElement}
   * @example
   * ```tsx
   * <Menu trigger={<button>Open Menu</button>}>
   *   // menu items
   * </Menu>
   * ```
   */
  trigger?: JSX.Element;
}

/**
 * Props for the MenuPopup component.
 *
 * The MenuPopup extends the base Menu component with additional positioning
 * and visibility controls. It handles the popup behavior and positioning
 * of the menu relative to its trigger element.
 *
 * @interface IMenuPopupProps
 */
export interface IMenuPopupProps extends IMenuProps {
  /**
   * Controls the horizontal alignment of the menu popup relative to its trigger element.
   * - 'left': Aligns the menu to the left edge of the trigger
   * - 'right': Aligns the menu to the right edge of the trigger
   *
   * @type {'left' | 'right'}
   */
  ali: 'left' | 'right';

  /**
   * Controls the visibility state of the menu popup.
   * When true, the menu is displayed; when false, it's hidden.
   * This is typically controlled by the parent component's state.
   *
   * @type {boolean}
   */
  isVisible: boolean;

  /**
   * Controls the vertical positioning of the menu popup relative to its trigger element.
   * - 'top': Positions the menu above the trigger element
   * - 'bottom': Positions the menu below the trigger element
   *
   * @type {'top' | 'bottom'}
   */
  pos: 'top' | 'bottom';
}

/**
 * Props for individual MenuItem components.
 *
 * The MenuItem component represents a single selectable item within a menu.
 * It typically rendered as list items for semantic correctness.
 *
 * @interface IMenuItemProps
 */
export interface IMenuItemProps extends IBaseProps, HTMLAttributes<HTMLLIElement> {
  /**
   * Optional icon element to display alongside the menu item text.
   * The icon will typically be positioned to the left of the text content.
   * Can be any JSX element, commonly an icon component or SVG.
   *
   * @type {JSX.Element}
   * @example
   * ```tsx
   * <MenuItem icon={<HomeIcon />}>Home</MenuItem>
   * <MenuItem icon={<SettingsIcon />}>Settings</MenuItem>
   * ```
   */
  icon?: JSX.Element;

  /**
   * Controls whether the menu item has sharp (squared) corners or rounded corners.
   * When true, the menu item will display with sharp corners instead of the default rounded appearance.
   * This should typically match the parent menu's isSharp setting for visual consistency.
   *
   * @type {boolean}
   * @default false
   */
  isSharp?: boolean;

  /**
   * The tone of the menu item, which determines the color and appearance.
   *
   * @type {TTone}
   */
  tone?: TTone;

  /**
   * The variant of the menu item, which determines the visual style.
   *
   * @type {'primary' | 'outlined' | 'simple'}
   */
  variant?: 'primary' | 'outlined' | 'simple';
}
