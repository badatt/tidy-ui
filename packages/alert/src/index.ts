export * from './Alert';
export * from './AlertFooter';
export * from './AlertTitle';
