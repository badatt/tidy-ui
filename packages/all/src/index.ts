export * from '@tidy-ui/commons';
export * from '@tidy-ui/feedback';
export * from '@tidy-ui/form';
export * from '@tidy-ui/layout';
export * from '@tidy-ui/navigation';
export * from '@tidy-ui/presentation';
