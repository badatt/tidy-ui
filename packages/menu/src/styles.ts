import { css, hsla } from '@tidy-ui/commons';
import { IMenuItemProps } from './types';

const simple = css<IMenuItemProps>`
  ${({ theme: { palette, isDark }, tone, disabled }) => css`
    color: ${isDark ? palette[tone!][400] : palette[tone!][800]};
    background-color: transparent;
    border: none;
    ${!disabled &&
    css`
      &:hover {
        background-color: ${isDark ? palette[tone!][900] : palette[tone!][200]};
      }
    `}
  `}
`;

const primary = css<IMenuItemProps>`
  ${({ theme: { palette, isDark }, tone, disabled }) => css`
    font-weight: 600;
    color: ${palette[tone!][50]};
    background-color: ${palette[tone!][700]};
    border: 1px solid ${hsla(palette[tone!].shades[900], 0.4)};
    ${!isDark &&
    css`
      box-shadow: 0px 1px 0px ${palette[tone!][100]};
    `}
    ${!disabled &&
    css`
      &:hover {
        background-color: ${palette[tone!][800]};
      }
    `}
  `}
`;

const outlined = css<IMenuItemProps>`
  ${({ theme: { palette, isDark }, tone, disabled }) => css`
    color: ${isDark ? palette[tone!][100] : palette[tone!][900]};
    border: 1px solid ${isDark ? palette[tone!][700] : palette[tone!][400]};
    ${!disabled &&
    css`
      &:hover {
        background-color: ${isDark ? hsla(palette[tone!].shades[800], 0.2) : palette[tone!][100]};
        border: 1px solid ${isDark ? palette[tone!][400] : palette[tone!][900]};
      }
    `}
  `}
`;

export { outlined, primary, simple };
