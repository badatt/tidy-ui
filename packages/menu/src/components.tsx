import { keyframes } from 'styled-components';
import { applyStandardOverrideStyles, color, css, hsla, styled } from '@tidy-ui/commons';
import { outlined, primary, simple } from './styles';
import { IMenuItemProps, IMenuPopupProps } from './types';

const fadeInScale = keyframes`
  from {
    opacity: 0;
    transform: scale(0.95);
  }
  to {
    opacity: 1;
    transform: scale(1);
  }
`;

const fadeOutScale = keyframes`
  from {
    opacity: 1;
    transform: scale(1);
  }
  to {
    opacity: 0;
    transform: scale(0.95);
  }
`;

const MenuWrapper = styled.div`
  position: relative;
  display: inline-block;
`;

/** @internal */
const MenuPopup = styled.menu<IMenuPopupProps>`
  position: absolute;
  z-index: 1000;
  display: flex;
  flex-direction: column;
  gap: 2px;
  padding: 0.5rem;
  ${({ theme: { palette, layout }, isSharp }) => css`
    background-color: ${palette.background.paper};
    border-radius: ${!isSharp && layout.radius};
    box-shadow: ${layout.shadow};
    animation: ${(isVisible) => (isVisible ? fadeInScale : fadeOutScale)} 0.15s ease-out forwards;
    transform-origin: top left;
  `}
  ${({ pos }) =>
    pos === 'top'
      ? css`
          bottom: 100%;
          margin-bottom: 0.5rem;
        `
      : css`
          top: 100%;
          margin-top: 0.5rem;
        `};

  ${({ ali }) =>
    ali === 'right'
      ? css`
          right: 0;
        `
      : css`
          left: 0;
        `};
  ${applyStandardOverrideStyles}
`;

const MenuItemRoot = styled.li<IMenuItemProps>`
  display: flex;
  padding: 0.5rem;
  align-items: center;
  cursor: pointer;
  ${({ theme: { layout }, isSharp, variant }) => css`
    border-radius: ${!isSharp && layout.radius};
    ${variant === 'simple' && simple}
    ${variant === 'primary' && primary}
    ${variant === 'outlined' && outlined}
  `}

  ${applyStandardOverrideStyles}
`;

const MenuItemIcon = styled.i<IMenuItemProps>`
  display: flex;
  align-items: center;
  padding-right: 0.5em;
`;

/** @internal */
const getShade = (isDark: boolean) => (isDark ? 700 : 300);

const MenuItemSeparatorRoot = styled.hr<IMenuItemProps>`
  height: 1px;
  margin: 0.5rem 0;
  ${({ theme: { isDark } }) => css`
    background-color: ${hsla(color.slate[getShade(isDark)])};
  `}
  ${applyStandardOverrideStyles}
  cursor: block;
`;

export { MenuItemIcon, MenuItemRoot, MenuItemSeparatorRoot, MenuPopup, MenuWrapper };
