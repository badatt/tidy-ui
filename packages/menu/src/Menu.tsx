import React from 'react';
import { MenuPopup, MenuTrigger, MenuWrapper } from './components';
import { IMenuProps } from './types';

const Menu = React.forwardRef<HTMLMenuElement, IMenuProps>((props, ref) => {
  const { children, ele, trigger, ...rest } = props;
  const menuRef = React.useRef<HTMLDivElement>(null);
  const triggerRef = React.useRef<HTMLDivElement>(null);

  const [open, setOpen] = React.useState(false);
  const [isVisible, setIsVisible] = React.useState(false);
  const [position, setPosition] = React.useState<'bottom' | 'top'>('bottom');
  const [align, setAlign] = React.useState<'left' | 'right'>('left');

  /** @internal */
  const toggleMenu = () => setOpen((prev) => !prev);

  /** @internal */
  const updatePosition = () => {
    if (!triggerRef.current || !menuRef.current) return;

    const triggerRect = triggerRef.current.getBoundingClientRect();
    const menuRect = menuRef.current.getBoundingClientRect();

    // Vertical position
    const shouldShowAbove = triggerRect.bottom + menuRect.height > window.innerHeight;
    setPosition(shouldShowAbove ? 'top' : 'bottom');

    // Horizontal alignment
    const shouldAlignRight = triggerRect.left + menuRect.width > window.innerWidth;
    setAlign(shouldAlignRight ? 'right' : 'left');
  };

  React.useEffect(() => {
    if (open) {
      setIsVisible(true);
    } else {
      const timeout = setTimeout(() => setIsVisible(false), 150);
      return () => clearTimeout(timeout);
    }
  }, [open]);

  React.useEffect(() => {
    /** @internal */
    const handleClickOutside = (e: MouseEvent) => {
      if (menuRef.current && !menuRef.current.contains(e.target as Node)) {
        setOpen(false);
      }
    };
    if (open) {
      document.addEventListener('mousedown', handleClickOutside);
    }
    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [open]);

  React.useEffect(() => {
    if (open) {
      updatePosition();
      window.addEventListener('resize', updatePosition);
      window.addEventListener('scroll', updatePosition);
    }
    return () => {
      window.removeEventListener('resize', updatePosition);
      window.removeEventListener('scroll', updatePosition);
    };
  }, [open]);

  /** @internal */
  const handleClick = (e) => {
    trigger.props.onClick?.(e);
    toggleMenu();
  };

  /** @internal */
  const renderChildren = (children): React.ReactNode => {
    return React.Children.map(children, (c) => {
      const child = c as React.ReactElement;
      return React.cloneElement(child, { isSharp: rest.isSharp });
    });
  };

  return (
    <MenuWrapper data-tui-name="MenuTrigger" ref={menuRef}>
      <MenuTrigger ref={triggerRef} onClick={handleClick}>
        {trigger}
      </MenuTrigger>
      {isVisible && (
        <MenuPopup role="menu" data-tui-name="Menu" isVisible={open} ref={ref} pos={position} ali={align} {...rest}>
          {ele ? React.cloneElement(ele, {}, renderChildren(children)) : renderChildren(children)}
        </MenuPopup>
      )}
    </MenuWrapper>
  );
});

Menu.defaultProps = {
  isSharp: false,
};

Menu.displayName = 'Menu';

export { Menu };
