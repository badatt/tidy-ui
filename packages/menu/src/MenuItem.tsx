import React from 'react';
import { MenuItemIcon, MenuItemRoot } from './components';
import { IMenuItemProps } from './types';

const MenuItem = React.forwardRef<HTMLLIElement, IMenuItemProps>((props, ref) => {
  const { children, ele, icon, ...rest } = props;
  return (
    <MenuItemRoot ref={ref} data-tui-name="MenuItem" role="menuitem" {...rest}>
      {icon && <MenuItemIcon>{icon}</MenuItemIcon>}
      {ele ? React.cloneElement(ele, {}, children) : children}
    </MenuItemRoot>
  );
});

MenuItem.defaultProps = {
  isSharp: false,
};

MenuItem.displayName = 'MenuItem';

export { MenuItem };
