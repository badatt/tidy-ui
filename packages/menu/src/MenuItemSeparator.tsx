import React from 'react';
import { MenuItemSeparatorRoot } from './components';

const MenuItemSeparator = React.forwardRef<HTMLHRElement>((props, ref) => {
  const { ...rest } = props;
  return <MenuItemSeparatorRoot ref={ref} data-tui-name="MenuItemSeparator" {...rest} />;
});

MenuItemSeparator.defaultProps = {};

MenuItemSeparator.displayName = 'MenuItemSeparator';

export { MenuItemSeparator };
