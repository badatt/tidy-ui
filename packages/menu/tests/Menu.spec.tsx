/**
 * @jest-environment jsdom
 */
import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import { orchidDark, orchidLight, TidyUiProvider } from '../../commons/src';
import { Menu, MenuItem, MenuItemSeparator } from '../src';

describe('Menu', () => {
  it('Basic render', () => {
    const tree = render(
      <TidyUiProvider theme={orchidLight}>
        <Menu trigger={<div data-testid="menu-btn">Menu</div>}>
          <MenuItem icon={<i>icon</i>}>Home</MenuItem>
          <MenuItemSeparator />
          <MenuItem>Logout</MenuItem>
        </Menu>
      </TidyUiProvider>,
    );
    const trigger = tree.getByTestId('menu-btn');
    fireEvent.click(trigger);
    fireEvent.click(tree.container);
    expect(tree).toMatchSnapshot();
  });
  it('Dark mode basic render', () => {
    const tree = render(
      <TidyUiProvider theme={orchidDark}>
        <Menu trigger={<div data-testid="menu-btn">Menu</div>}>
          <MenuItem icon={<i>icon</i>}>Home</MenuItem>
          <MenuItemSeparator />
          <MenuItem>Logout</MenuItem>
        </Menu>
      </TidyUiProvider>,
    );
    const trigger = tree.getByTestId('menu-btn');
    fireEvent.click(trigger);
    fireEvent.click(tree.container);
    expect(tree).toMatchSnapshot();
  });
});
