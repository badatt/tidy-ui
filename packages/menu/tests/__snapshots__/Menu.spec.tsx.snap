// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Basic render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  position: relative;
  display: inline-block;
}

.c2 {
  position: absolute;
  z-index: 1000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0.5rem;
  background-color: hsla(0,100%,100%,1);
  border-radius: 0.5rem;
  box-shadow: rgb(255,255,255) 0px 0px 0px 0px,rgba(15,23,42,0.05) 0px 0px 0px 1px,rgba(0,0,0,0.1) 0px 20px 25px -5px,rgba(0,0,0,0.1) 0px 8px 10px -6px;
  -webkit-animation: ikRgWa 0.15s ease-out forwards;
  animation: ikRgWa 0.15s ease-out forwards;
  -webkit-transform-origin: top left;
  -ms-transform-origin: top left;
  transform-origin: top left;
  top: 100%;
  margin-top: 0.5rem;
  left: 0;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  border-radius: 0.5rem;
}

.c3:hover {
  background-color: hsla(213,27%,84%,0.5);
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-right: 0.5em;
}

.c5 {
  height: 1px;
  margin: 0.5rem 0;
  background-color: hsla(213,27%,84%,1);
}

.c1 {
  cursor: pointer;
}

<body>
    <div>
      <div
        class="c0"
        data-tui-name="MenuTrigger"
      >
        <div
          class="c1"
        >
          <div
            data-testid="menu-btn"
          >
            Menu
          </div>
        </div>
        <menu
          class="c2"
          data-tui-name="Menu"
          role="menu"
        >
          <li
            class="c3"
            data-tui-name="MenuItem"
            role="menuitem"
          >
            <i
              class="c4"
            >
              <i>
                icon
              </i>
            </i>
            Home
          </li>
          <hr
            class="c5"
            data-tui-name="MenuItemSeparator"
          />
          <li
            class="c3"
            data-tui-name="MenuItem"
            role="menuitem"
          >
            Logout
          </li>
        </menu>
      </div>
    </div>
  </body>,
  "container": .c0 {
  position: relative;
  display: inline-block;
}

.c2 {
  position: absolute;
  z-index: 1000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0.5rem;
  background-color: hsla(0,100%,100%,1);
  border-radius: 0.5rem;
  box-shadow: rgb(255,255,255) 0px 0px 0px 0px,rgba(15,23,42,0.05) 0px 0px 0px 1px,rgba(0,0,0,0.1) 0px 20px 25px -5px,rgba(0,0,0,0.1) 0px 8px 10px -6px;
  -webkit-animation: ikRgWa 0.15s ease-out forwards;
  animation: ikRgWa 0.15s ease-out forwards;
  -webkit-transform-origin: top left;
  -ms-transform-origin: top left;
  transform-origin: top left;
  top: 100%;
  margin-top: 0.5rem;
  left: 0;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  border-radius: 0.5rem;
}

.c3:hover {
  background-color: hsla(213,27%,84%,0.5);
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-right: 0.5em;
}

.c5 {
  height: 1px;
  margin: 0.5rem 0;
  background-color: hsla(213,27%,84%,1);
}

.c1 {
  cursor: pointer;
}

<div>
    <div
      class="c0"
      data-tui-name="MenuTrigger"
    >
      <div
        class="c1"
      >
        <div
          data-testid="menu-btn"
        >
          Menu
        </div>
      </div>
      <menu
        class="c2"
        data-tui-name="Menu"
        role="menu"
      >
        <li
          class="c3"
          data-tui-name="MenuItem"
          role="menuitem"
        >
          <i
            class="c4"
          >
            <i>
              icon
            </i>
          </i>
          Home
        </li>
        <hr
          class="c5"
          data-tui-name="MenuItemSeparator"
        />
        <li
          class="c3"
          data-tui-name="MenuItem"
          role="menuitem"
        >
          Logout
        </li>
      </menu>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Menu Dark mode basic render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  position: relative;
  display: inline-block;
}

.c2 {
  position: absolute;
  z-index: 1000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0.5rem;
  background-color: hsla(214,13%,20%,1);
  border-radius: 0.5rem;
  box-shadow: rgb(255,255,255) 0px 0px 0px 0px,rgba(15,23,42,0.05) 0px 0px 0px 1px,rgba(0,0,0,0.1) 0px 20px 25px -5px,rgba(0,0,0,0.1) 0px 8px 10px -6px;
  -webkit-animation: ikRgWa 0.15s ease-out forwards;
  animation: ikRgWa 0.15s ease-out forwards;
  -webkit-transform-origin: top left;
  -ms-transform-origin: top left;
  transform-origin: top left;
  top: 100%;
  margin-top: 0.5rem;
  left: 0;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  border-radius: 0.5rem;
}

.c3:hover {
  background-color: hsla(215,25%,27%,0.5);
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-right: 0.5em;
}

.c5 {
  height: 1px;
  margin: 0.5rem 0;
  background-color: hsla(215,25%,27%,1);
}

.c1 {
  cursor: pointer;
}

<body>
    <div>
      <div
        class="c0"
        data-tui-name="MenuTrigger"
      >
        <div
          class="c1"
        >
          <div
            data-testid="menu-btn"
          >
            Menu
          </div>
        </div>
        <menu
          class="c2"
          data-tui-name="Menu"
          role="menu"
        >
          <li
            class="c3"
            data-tui-name="MenuItem"
            role="menuitem"
          >
            <i
              class="c4"
            >
              <i>
                icon
              </i>
            </i>
            Home
          </li>
          <hr
            class="c5"
            data-tui-name="MenuItemSeparator"
          />
          <li
            class="c3"
            data-tui-name="MenuItem"
            role="menuitem"
          >
            Logout
          </li>
        </menu>
      </div>
    </div>
  </body>,
  "container": .c0 {
  position: relative;
  display: inline-block;
}

.c2 {
  position: absolute;
  z-index: 1000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0.5rem;
  background-color: hsla(214,13%,20%,1);
  border-radius: 0.5rem;
  box-shadow: rgb(255,255,255) 0px 0px 0px 0px,rgba(15,23,42,0.05) 0px 0px 0px 1px,rgba(0,0,0,0.1) 0px 20px 25px -5px,rgba(0,0,0,0.1) 0px 8px 10px -6px;
  -webkit-animation: ikRgWa 0.15s ease-out forwards;
  animation: ikRgWa 0.15s ease-out forwards;
  -webkit-transform-origin: top left;
  -ms-transform-origin: top left;
  transform-origin: top left;
  top: 100%;
  margin-top: 0.5rem;
  left: 0;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  border-radius: 0.5rem;
}

.c3:hover {
  background-color: hsla(215,25%,27%,0.5);
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-right: 0.5em;
}

.c5 {
  height: 1px;
  margin: 0.5rem 0;
  background-color: hsla(215,25%,27%,1);
}

.c1 {
  cursor: pointer;
}

<div>
    <div
      class="c0"
      data-tui-name="MenuTrigger"
    >
      <div
        class="c1"
      >
        <div
          data-testid="menu-btn"
        >
          Menu
        </div>
      </div>
      <menu
        class="c2"
        data-tui-name="Menu"
        role="menu"
      >
        <li
          class="c3"
          data-tui-name="MenuItem"
          role="menuitem"
        >
          <i
            class="c4"
          >
            <i>
              icon
            </i>
          </i>
          Home
        </li>
        <hr
          class="c5"
          data-tui-name="MenuItemSeparator"
        />
        <li
          class="c3"
          data-tui-name="MenuItem"
          role="menuitem"
        >
          Logout
        </li>
      </menu>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
